### ISSUE TITLE
Wrong behaviour of HTTP -> HTTPS redirect with request-modifier (ReplacePathRegex)

### ISSUE THREAD

**Issue by kachkaev** [containous/traefik#3741]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

There's been a long-standing issue with preserving paths when redirecting from HTTP to HTTPS (https://github.com/containous/traefik/issues/1957), but it was finally fixed in https://github.com/containous/traefik/pull/3631 by @dtomcej. This is great stuff and I can confirm that `PathPrefixStrip` works properly in `1.7.3-rc3` 🎉 

However, it looks like the problem is still out there when we use `ReplacePathRegex`.

Here's an ingress annotation that works correctly:

```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: example-ingress
  annotations:
    traefik.ingress.kubernetes.io/rule-type: PathPrefixStrip
spec:
  rules:
  - host: example.com
    http:
      paths:
      - path: /backend-path
        backend:
          serviceName: example-backend
          servicePort: http
```

```bash
http http://example.com/backend-path/abc
# HTTP/1.1 302 Found
# Content-Length: 5
# Content-Type: text/plain; charset=utf-8
# Date: Mon, 06 Aug 2018 20:55:37 GMT
# Location: https://example.com/backend-path/abc
# 
# Found
```

And here's the same rule idea, but implemented via `request-modifier` / `ReplacePathRegex`:

```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: example-ingress
  annotations:
    traefik.ingress.kubernetes.io/request-modifier: "ReplacePathRegex: ^/backend-path-2/(.*) /$1"
spec:
  rules:
  - host: example.com
    http:
      paths:
      - path: /backend-path-2
        backend:
          serviceName: example-backend
          servicePort: http
```

```bash
http http://example.com/backend-path-2/abc
# HTTP/1.1 302 Found
# Content-Length: 5
# Content-Type: text/plain; charset=utf-8
# Date: Mon, 06 Aug 2018 20:55:40 GMT
# Location: https://example.com/abc
# 
# Found
```

### What did you expect to see?

I expected `"ReplacePathRegex: ^/backend-path-2/(.*) /$1"` work just like `PathPrefixStrip`, but it acted as before https://github.com/containous/traefik/pull/3631.

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
official helm chart, traefik image tag: `1.7.0-rc3-alpine`
```

---
**PS**: I switched to `ReplacePathRegexp` in an attempt to configure both path prefix stripping and adding trailing slash in one ingress rule (see [my stackoveflow question](https://stackoverflow.com/questions/51697574/adding-trailing-slash-to-microservice-root-path-in-traefik-ingress)). Not sure if it's the best solution though, regardless of this issue with HTTP → HTTPS redirects.

**Comment by kachkaev** [comment]
Here is one more discovery:

```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: example-ingress
  namespace: sb
  annotations:
    traefik.frontend.rule.type: PathPrefixStrip
spec:
  rules:
  - host: example.com
    http:
      paths:
      - path: /backend-path/ # <--------------------------------- trailing slash
        backend:
          serviceName: example-backend
          servicePort: http
```

```bash
http http://example.com/backend-path/abc
# HTTP/1.1 302 Found
# Content-Length: 5
# Content-Type: text/plain; charset=utf-8
Date: Tue, 07 Aug 2018 07:13:46 GMT
# Location: https://example.com/backend-path//abc   # <---------- double slash
# 
# Found
```

**Comment by JPM84** [comment]
I think I'm experiencing a similar issue also with 1.7.0-rc3-alpine.
Config is as follows:
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: my-frontend
  annotations:
    kubernetes.io/ingress.class: traefik
    traefik.ingress.kubernetes.io/frontend-entry-points: http,https
    traefik.ingress.kubernetes.io/redirect-entry-point: https
    traefik.ingress.kubernetes.io/request-modifier: ReplacePathRegex:^/(.*) /pages/$1
spec:
  rules:
  - host: mydomain.com
    http:
      paths:
      - path: /
        backend:
          serviceName: backend
          servicePort: backend
```

**Expected redirect bahaviour:**
http://mydomain.com/abc -> https://mydomain.com/abc

**Found redirect result:**
http://mydomain.com/abc -> https://mydomain.com/pages/abc

**Comment by dtomcej** [comment]
@kachkaev As I am reviewing this issue, Just wanted to point out:
`ReplacePathRegex: ^/backend-path-2/(.*) /$1`
and
`PathPrefixStrip: /backend-path-2` are not the same 😉   But I know what you are trying to say 😜

**Comment by kachkaev** [comment]
Is the difference only in that `PathPrefixStrip: /backend-path-2` does not include a trailing slash? Or am I missing something else?

**Comment by dtomcej** [comment]
Correct. The first will not match a request with a trailing slash, and the second will match both.

Similarly, when you add a trailing slash on the rule, the equivalent regex would require a lookbehind to determine whether there is a slash there 😛

**Comment by garyfritz** [comment]
While the alternative suggested to @JPM84's works in that case, however, replacing `ReplacePathRegex` by a combination of `PathPrefixStrip` and `AddPrefix` doesn't work (yet?).

Altering @kachkaev's intention:
```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: example-ingress
  annotations:
    # intended behaviour:
    # traefik.ingress.kubernetes.io/request-modifier: "ReplacePathRegex: ^/frontend-path/(.*) /backend-path/$1"

    # intended/naive substitution:
    traefik.ingress.kubernetes.io/request-modifier: "PathPrefixStrip:/frontend-path;AddPrefix:/backend-path"

spec:
  rules:
  - host: example.com
    http:
      paths:
      - path: /frontend-path
        backend:
          serviceName: example-backend
          servicePort: http
```
While the [documentation on their order](https://docs.traefik.io/v1.7/basics/#rules-order) claims that `PathPrefixStrip` was **both** and *matcher* and a *modifier*, Traefik acting according to its current implementation in the [`parseRequestModifier`](https://github.com/containous/traefik/pull/3481/files#diff-ebf930b0f073b06115d5f6c5172ea92eR578) method and thus complains:

> Failed to get rule for ingress default/my-frontend: cannot use non-modifier rule: "PathPrefixStripRegex"

@dtomcej Would you consider this a new bug or is this concerning the code you're working on anyways (or both)?

**Comment by dtomcej** [comment]
@garyfritz the `Path(x)` rules are not both a matcher and a modifier. They are matchers only. They are shown together in the list of orders to demonstrate that Matchers will always precede modifiers.

To replace a current replace path regex, you should set `traefik.ingress.kubernetes.io/rule-type: PathPrefixStrip`, and then use `traefik.ingress.kubernetes.io/request-modifier: AddPrefix:/backend-path`

This is not a bug, and is not linked to the current ticket.

**Comment by garyfritz** [comment]
Many thanks @dtomcej, this indeed solves the outlined case and is by far less complicated than imagined :wink: Looks like I got confused/mislead by the wording in the [documentation](https://docs.traefik.io/v1.7/basics/#rules-order).

**Comment by dtomcej** [comment]
@garyfritz No worries :) I should have the replacepath(regex) with redirects working shortly, and a much improved test suite to ensure they stay that way 😄

**Comment by traefiker** [comment]
Closed by #3742.
